Bluetooth: audio: BAP: allow for multiple callbacks

kruithofa opened the pull request:
The existing implementation allows for only one callback per procedure. This creates problems when for example adding CAP functionality. The callbacks are here implemented as an SLIST

Commit:
Bluetooth: audio: BAP: allow for multiple callbacks

The existing implementation allows for only one callback per procedure.
This creates problems when for example adding CAP functionality.
The callbacks are here implemented as an SLIST

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

Thalley commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -381,11 +383,18 @@ static uint8_t notify_handler(struct bt_conn *conn,
+		SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&broadcast_assistant_cbs,
+						  listener, next, _node) {
+			if (listener->recv_state_removed) {
+				broadcast_assistant.past_avail[index] = false;
```
This looks like a bug: `broadcast_assistant.past_avail[index] = false;` is only done if there is a callback, but it should always be done, right? Suggest to fix in this PR, but as a separate commit

Thalley commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -611,10 +640,11 @@ static uint8_t service_discover_func(struct bt_conn *conn,
-			if (broadcast_assistant_cbs != NULL &&
-			    broadcast_assistant_cbs->discover != NULL) {
-				broadcast_assistant_cbs->discover(conn, err, 0);
+			SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&broadcast_assistant_cbs,
+							  listener, next, _node) {
+				if (listener->discover) {
+					listener->discover(conn, err, 0);
+				}
```
Suggest to create helper functions for calling the callback, as we've done for e.g. VCP and MICP: 
```c
static void vcp_vol_ctlr_flags_changed(struct bt_vcp_vol_ctlr *vol_ctlr, int err)
{
	struct bt_vcp_vol_ctlr_cb *listener, *next;

	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&vcp_vol_ctlr_cbs, listener, next, _node) {
		if (listener->flags) {
			listener->flags(vol_ctlr, err, err == 0 ? vol_ctlr->flags : 0);
		}
	}
}
```

This applies for all callbacks

kruithofa replied:
I didn't update the ones in the bap_broadcast_assistant_write_cp_cb, these are called only once. This is similar to what was done for VCP

Thalley commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -861,9 +892,40 @@ int bt_bap_broadcast_assistant_discover(struct bt_conn *conn)
+		}
+	}
+
+	LOG_DBG("Adding callbacks to the list");
```
```suggestion
```

Thalley reviewed (changes requested):
Some suggestions, but otherwise looks good

Commit:
Bluetooth: audio: BAP: set the past_avail status

While refactoring the handling of callbacks the setting of past_avail
status got deleted (and it was in the wrong place). Adding it back at
the correct place.

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

Thalley commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -85,6 +85,57 @@ static int16_t lookup_index_by_handle(uint16_t handle)
+	}
+}
+
+static void bap_broadcast_assistant_recv_state(struct bt_conn *conn, int err,
+					       const struct bt_bap_scan_delegator_recv_state *s)
```
```suggestion
static void bap_broadcast_assistant_recv_state_changed(struct bt_conn *conn, int err,
					       const struct bt_bap_scan_delegator_recv_state *s)
```

kruithofa replied:
That's what I had first, but then it came over the 100 char limit. So need to choose what is the least bad here...

Thalley reviewed (changes requested):
One change request, and a few optional suggestions

kruithofa commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -85,6 +85,57 @@ static int16_t lookup_index_by_handle(uint16_t handle)
+}
+
+static void bap_broadcast_assistant_recv_state(struct bt_conn *conn, int err,
+					       const struct bt_bap_scan_delegator_recv_state *s)
+{
+	struct bt_bap_broadcast_assistant_cb *listener, *next;
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&broadcast_assistant_cbs,
+					  listener, next, _node) {
+		if (listener->recv_state) {
+			listener->recv_state(conn, err, s);
```
I know but there is the 100 char limit

kruithofa commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -85,6 +85,57 @@ static int16_t lookup_index_by_handle(uint16_t handle)
+	}
+}
+
+static void bap_broadcast_assistant_recv_state_removed(struct bt_conn *conn, int err,
+						       uint8_t src_id)
+{
+	struct bt_bap_broadcast_assistant_cb *listener, *next;
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&broadcast_assistant_cbs,
+					  listener, next, _node) {
+		if (listener->recv_state_removed) {
+			listener->recv_state_removed(conn, err, src_id);
+		}
+	}
+}
```
That's an API change right? Maybe a separate PR

Thalley replied:
https://github.com/zephyrproject-rtos/zephyr/issues/69527

kruithofa commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -85,6 +85,57 @@ static int16_t lookup_index_by_handle(uint16_t handle)
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&broadcast_assistant_cbs,
+					  listener, next, _node) {
+		if (listener->discover) {
+			listener->discover(conn, err, recv_state_count);
+		}
```
when `err!=0` the caller already sets `recv_state_count` to 0

Thalley commented on subsys/bluetooth/audio/bap_broadcast_assistant.c:
```diff
@@ -85,6 +85,58 @@ static int16_t lookup_index_by_handle(uint16_t handle)
+	}
+}
+
+static void bap_broadcast_assistant_recv_state_changed(struct bt_conn *conn, int err,
+						       const struct bt_bap_scan_delegator_recv_state
+						       *state)
```
```suggestion
static void
bap_broadcast_assistant_recv_state_changed(struct bt_conn *conn, int err,
					   const struct bt_bap_scan_delegator_recv_state *state)
```

or
```suggestion
static void bap_broadcast_assistant_recv_state_changed(
	struct bt_conn *conn, int err, const struct bt_bap_scan_delegator_recv_state *state)
```
would be better than adding a newline between the type and variable (consider using clang-format to help format)

Thalley reviewed (changes requested):
Minor formatting request
